engines/asylum/views/scene.cpp

```diff
@@ -0,0 +1,2886 @@
+
+		// Enable player
+		getActor()->updateStatus(kActorStatusEnabled);
+	}
+
+	uint32 ticks = _vm->getTick();
+
+	if (!getSharedData()->getFlag(kFlagRedraw)) {
+		if (updateScreen())
+			return true;
+
+		getSharedData()->setFlag(kFlagRedraw, true);
+	}
+
+	if (ticks > getSharedData()->getNextScreenUpdate()) {
+		if (getSharedData()->getFlag(kFlagRedraw)) {
+			if (getSharedData()->getMatteBarHeight() <= 0)
+				getScreen()->copyBackBufferToScreen();
+			else
+				getEncounter()->drawScreen();
+
+			// Original also sets an unused value to 0
+			getSharedData()->setEventUpdate(getSharedData()->getEventUpdate() ^ 1);
+
+			getSharedData()->setFlag(kFlagRedraw, false);
+			getSharedData()->setNextScreenUpdate(ticks + 55);
+			++_vm->screenUpdateCount;
+		}
+	}
+
+	return true;
+}
+
+bool Scene::key(const AsylumEvent &evt) {
+	if (!_ws)
+		error("[Scene::key] WorldStats not initialized properly");
+
+	// TODO add support for debug commands
+
+	//////////////////////////////////////////////////////////////////////////
+	// Check for keyboard shortcuts
+	if (evt.kbd.ascii == Config.keyShowVersion) {
+		// TODO show version!
+
+		return true;
+	}
+
+	if (evt.kbd.ascii == Config.keyQuickLoad) {
+		getSaveLoad()->quickLoad();
+
+		return true;
+	}
+
+	if (evt.kbd.ascii == Config.keyQuickSave) {
+		getSaveLoad()->quickSave();
+
+		return true;
+	}
+
+	if (evt.kbd.ascii == Config.keySwitchToSara) {
+		if (getCursor()->isHidden() || _ws->chapter != kChapter9)
+			return true;
+
+		getScript()->queueScript(_ws->actions[_ws->getActionAreaIndexById(2206)]->scriptIndex, getSharedData()->getPlayerIndex());
+
+		return true;
+	}
+
+	if (evt.kbd.ascii == Config.keySwitchToGrimwall) {
+		if (getCursor()->isHidden() || _ws->chapter != kChapter9)
+			return true;
+
+		getScript()->queueScript(_ws->actions[_ws->getActionAreaIndexById(2207)]->scriptIndex, getSharedData()->getPlayerIndex());
+
+		return true;
+	}
+
+	if (evt.kbd.ascii == Config.keySwitchToOlmec) {
+		if (getCursor()->isHidden() || _ws->chapter != kChapter9)
+			return true;
+
+		getScript()->queueScript(_ws->actions[_ws->getActionAreaIndexById(2208)]->scriptIndex, getSharedData()->getPlayerIndex());
+
+		return true;
+	}
+
+	switch (evt.kbd.keycode) {
+	default:
+		break;
+
+	case Common::KEYCODE_BACKSPACE:
+		// TODO add support for debug commands
+		warning("[Scene::key] debug command handling not implemented!");
+		break;
+
+	case Common::KEYCODE_RETURN:
+		// TODO add support for debug commands
+		warning("[Scene::key] debug command handling not implemented!");
+		break;
+
+	case Common::KEYCODE_ESCAPE:
+		// TODO add support for debug commands
+
+		if (getSpeech()->getSoundResourceId()) {
+			getScene()->stopSpeech();
+		} else {
+			if (getCursor()->isHidden())
+				break;
+
+			_vm->switchEventHandler(_vm->menu());
+		}
+		break;
+
+	case Common::KEYCODE_LEFTBRACKET:
+		if (evt.kbd.ascii != 123)
+			break;
+		// Fallback to next case (we got a left brace)
```

Comment:
Compiler Warning:
````
engines/asylum/views/scene.cpp: In member function ‘bool Asylum::Scene::key(const Asylum::AsylumEvent&)’:
engines/asylum/views/scene.cpp:492:3: warning: this statement may fall through [-Wimplicit-fallthrough=]
   if (evt.kbd.ascii != 123)
   ^~
engines/asylum/views/scene.cpp:496:2: note: here
  case Common::KEYCODE_p:
  ^~~~
  
  You need a comment similar to // fallthrough to nudge the compiler not to throw a warning since this does appear intentional.
````